File: public_web/tests/vixrex-dogal-netlestirme-pipeline.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";

const { rpcMock } = vi.hoisted(() => ({ rpcMock: vi.fn() }));

vi.mock("../src/lib/supabase", () => ({
  supabase: { rpc: rpcMock },
}));

import { handleVixrexNluMessage } from "../src/lib/vixrexNluPipeline";

function pending(
  anahtar: string,
  etiket: string,
  tip: string,
  eylem?: "kaldir",
) {
  rpcMock.mockImplementation(async (name: string, params?: unknown) => {
    if (name === "get_assistant_pending_slot") {
      return { data: { anahtar, etiket, tip, ...(eylem ? { eylem } : {}) }, error: null };
    }
    if (name === "set_assistant_pending_slot") {
      return { data: params ?? null, error: null };
    }
    return { data: null, error: null };
  });
}

describe("Vixrex doğal netleştirme gerçek Next pipeline", () => {
  beforeEach(() => {
    rpcMock.mockReset();
  });

  it("bağlam yok ve cümle anlaşılmıyorsa teknik alan adı sormaz", async () => {
    rpcMock.mockResolvedValue({ data: null, error: null });
    const sonuc = await handleVixrexNluMessage("bunu farklı yap");
    expect(sonuc.outcome).toBe("notUnderstood");
    expect(sonuc.message).toBe("Vitrininde neyi farklı görmek istersin?");
    expect(sonuc.message.toLocaleLowerCase("tr-TR")).not.toContain("hangi alan");
  });

  it("bağlam yokken 'telefonu değiştirme' yazma isteği sayılmaz", async () => {
    rpcMock.mockResolvedValue({ data: null, error: null });
    const sonuc = await handleVixrexNluMessage("telefonu değiştirme");
    expect(sonuc.outcome).toBe("needsClarification");
    expect(sonuc.tumu).toBeUndefined();
    expect(sonuc.message).toContain("değişiklik yapmıyorum");
  });

  it("telefon beklenirken 'değiştirme' işlemi iptal eder ve bekleyen soruyu temizler", async () => {
    pending("telefon", "Arama Numarası", "telefon");
    const sonuc = await handleVixrexNluMessage("değiştirme");
    expect(sonuc.outcome).toBe("needsClarification");
    expect(sonuc.tumu).toBeUndefined();
    expect(sonuc.message).toContain("bu değişikliği yapmıyorum");
    const clearCall = rpcMock.mock.calls.find(
      ([name, args]) => name === "set_assistant_pending_slot" && (args as { p_slot?: unknown })?.p_slot === null,
    );
    expect(clearCall).toBeTruthy();
  });

  it("telefon beklenirken 'evet' cevabını telefon değeri diye yazmaz ve bağlamı silmez", async () => {
    pending("telefon", "Arama Numarası", "telefon");
    const sonuc = await handleVixrexNluMessage("evet");
    expect(sonuc.outcome).toBe("needsClarification");
    expect(sonuc.anahtar).toBe("telefon");
    expect(sonuc.message).toBe("Arama Numarası için ne yazayım?");
    const clearCalls = rpcMock.mock.calls.filter(
      ([name, args]) => name === "set_assistant_pending_slot" && (args as { p_slot?: unknown })?.p_slot === null,
    );
    expect(clearCalls).toHaveLength(0);
  });

  it("aç/kapa sorusunda 'evet' cevabını önceki soruyla birlikte true olarak çözer", async () => {
    pending("puanGoster", "Puan Görünsün", "acikKapali");
    const sonuc = await handleVixrexNluMessage("evet");
    expect(sonuc.outcome).toBe("handled");
    expect(sonuc.anahtar).toBe("puanGoster");
    expect(sonuc.deger).toBe(true);
  });

  it("'onu kaldır' cevabında kaldırma onayı bağlamını saklar", async () => {
    pending("telefon", "Arama Numarası", "telefon");
    const sonuc = await handleVixrexNluMessage("onu kaldır");
    expect(sonuc.outcome).toBe("needsClarification");
    expect(sonuc.message).toBe("Arama Numarası bilgisini kaldırmamı mı istiyorsun?");
    const saveCall = rpcMock.mock.calls.find(
      ([name, args]) =>
        name === "set_assistant_pending_slot" &&
        (args as { p_slot?: { eylem?: string } })?.p_slot?.eylem === "kaldir",
    );
    expect(saveCall).toBeTruthy();
  });

  it("kaldırma onayı beklenirken 'evet' alanı temizleme sonucuna dönüşür", async () => {
    pending("telefon", "Arama Numarası", "telefon", "kaldir");
    const sonuc = await handleVixrexNluMessage("evet");
    expect(sonuc.outcome).toBe("handled");
    expect(sonuc.anahtar).toBe("telefon");
    expect(sonuc.deger).toBeNull();
    expect(sonuc.tumu).toEqual([{ anahtar: "telefon", kolon: "phone", deger: null }]);
    expect(sonuc.message).toBe("Arama Numarası bilgisini kaldırdım.");
  });

  it("çalışma saati beklenirken yalnız kapanış saati verilirse tam saat diye kaydetmez", async () => {
    pending("calismaSaatleri", "Açılış Saatleri", "metin");
    const sonuc = await handleVixrexNluMessage("akşam yedi");
    expect(sonuc.outcome).toBe("needsClarification");
    expect(sonuc.message).toContain("kaçta açıp kaçta kapandığınızı");
    expect(sonuc.tumu).toBeUndefined();
  });

  it("açık adres beklenirken yalnız ilçe verilirse tam adres diye kaydetmez", async () => {
    pending("adres", "İşletme Adresi", "uzunMetin");
    const sonuc = await handleVixrexNluMessage("Bağcılar");
    expect(sonuc.outcome).toBe("needsClarification");
    expect(sonuc.message).toBe("Açık adresi biraz daha ayrıntılı yazar mısın?");
    expect(sonuc.tumu).toBeUndefined();
  });
});

File: test/vixrex_dogal_netlestirme_pipeline_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:vixrex/config/vixrex_niyet_sozlugu.g.dart';
import 'package:vixrex/services/vixrex_nlu/vixrex_canonical_draft_writer.dart';
import 'package:vixrex/services/vixrex_nlu/vixrex_conversation_memory.dart';
import 'package:vixrex/services/vixrex_nlu/vixrex_nlu_pipeline.dart';

class _FakeMemory implements VixrexConversationMemoryPort {
  _FakeMemory(this.slot);

  VixrexPendingSlot? slot;
  int clearCount = 0;

  @override
  Future<VixrexPendingSlot?> loadPendingSlot({String? scope}) async => slot;

  @override
  Future<void> savePendingSlot(VixrexPendingSlot slot, {String? scope}) async {
    this.slot = slot;
  }

  @override
  Future<void> clearPendingSlot({String? scope}) async {
    clearCount += 1;
    slot = null;
  }
}

class _FakeWriter extends VixrexCanonicalDraftWriter {
  _FakeWriter();

  int writeCount = 0;
  List<Object?>? lastValues;

  @override
  Future<VixrexCanonicalWriteResult> write({
    required List<VixrexNiyetAlan> alanlar,
    required List<Object?> degerler,
  }) async {
    writeCount += 1;
    lastValues = degerler;
    return const VixrexCanonicalWriteResult(
      VixrexCanonicalWriteState.written,
      commandId: '00000000-0000-4000-8000-000000000001',
    );
  }
}

VixrexPendingSlot _slot(
  String anahtar,
  String etiket,
  String tip, {
  String? eylem,
}) {
  return VixrexPendingSlot(
    anahtar: anahtar,
    etiket: etiket,
    tip: tip,
    sorulduAt: DateTime(2026, 9, 10),
    eylem: eylem,
  );
}

Future<({bool ok, String? hata, Object? normalizedDeger})> _validate(
  VixrexNiyetAlan alan,
  String hamDeger,
) async {
  if (hamDeger.isEmpty) {
    return (ok: true, hata: null, normalizedDeger: null);
  }
  if (alan.tip == 'acikKapali') {
    if (hamDeger == 'true') {
      return (ok: true, hata: null, normalizedDeger: true);
    }
    if (hamDeger == 'false') {
      return (ok: true, hata: null, normalizedDeger: false);
    }
  }
  return (ok: true, hata: null, normalizedDeger: hamDeger);
}

void main() {
  group('Vixrex doğal netleştirme gerçek Flutter pipeline', () {
    test('bağlam yok ve cümle anlaşılmıyorsa teknik alan adı sormaz', () async {
      final memory = _FakeMemory(null);
      final writer = _FakeWriter();
      final pipeline = VixrexNluPipeline(
        memory: memory,
        canonicalWriter: writer,
      );

      final sonuc = await pipeline.handle(
        input: 'bunu farklı yap',
        controller: null,
        onValidate: _validate,
      );

      expect(sonuc.outcome, VixrexNluPipelineOutcome.notUnderstood);
      expect(sonuc.message.text, 'Vitrininde neyi farklı görmek istersin?');
      expect(sonuc.message.text.toLowerCase(), isNot(contains('hangi alan')));
      expect(writer.writeCount, 0);
    });

    test("bağlam yokken 'telefonu değiştirme' yazma isteği sayılmaz", () async {
      final memory = _FakeMemory(null);
      final writer = _FakeWriter();
      final pipeline = VixrexNluPipeline(
        memory: memory,
        canonicalWriter: writer,
      );

      final sonuc = await pipeline.handle(
        input: 'telefonu değiştirme',
        controller: null,
        onValidate: _validate,
      );

      expect(sonuc.outcome, VixrexNluPipelineOutcome.needsClarification);
      expect(sonuc.message.text, contains('değişiklik yapmıyorum'));
      expect(writer.writeCount, 0);
    });

    test(
      "telefon beklenirken 'değiştirme' işlemi iptal eder ve bekleyen soruyu temizler",
      () async {
        final memory = _FakeMemory(_slot('telefon', 'Telefon', 'telefon'));
        final writer = _FakeWriter();
        final pipeline = VixrexNluPipeline(
          memory: memory,
          canonicalWriter: writer,
        );

        final sonuc = await pipeline.handle(
          input: 'değiştirme',
          controller: null,
          onValidate: _validate,
        );

        expect(sonuc.outcome, VixrexNluPipelineOutcome.needsClarification);
        expect(sonuc.message.text, contains('bu değişikliği yapmıyorum'));
        expect(memory.clearCount, 1);
        expect(memory.slot, isNull);
        expect(writer.writeCount, 0);
      },
    );

    test(
      "telefon beklenirken 'evet' telefon değeri diye yazılmaz ve bağlam silinmez",
      () async {
        final memory = _FakeMemory(_slot('telefon', 'Telefon', 'telefon'));
        final writer = _FakeWriter();
        final pipeline = VixrexNluPipeline(
          memory: memory,
          canonicalWriter: writer,
        );

        final sonuc = await pipeline.handle(
          input: 'evet',
          controller: null,
          onValidate: _validate,
        );

        expect(sonuc.outcome, VixrexNluPipelineOutcome.needsClarification);
        expect(sonuc.appliedAnahtar, 'telefon');
        expect(sonuc.message.text, 'Telefon için ne yazayım?');
        expect(memory.clearCount, 0);
        expect(writer.writeCount, 0);
      },
    );

    test(
      "aç/kapa sorusunda 'evet' önceki soruyla true olarak çözülür",
      () async {
        final memory = _FakeMemory(
          _slot('puanGoster', 'Değerlendirme Puanını Göster', 'acikKapali'),
        );
        final writer = _FakeWriter();
        final pipeline = VixrexNluPipeline(
          memory: memory,
          canonicalWriter: writer,
        );

        final sonuc = await pipeline.handle(
          input: 'evet',
          controller: null,
          onValidate: _validate,
        );

        expect(sonuc.outcome, VixrexNluPipelineOutcome.handled);
        expect(sonuc.appliedAnahtar, 'puanGoster');
        expect(sonuc.appliedDeger, true);
        expect(memory.clearCount, 1);
        expect(writer.writeCount, 1);
      },
    );

    test("'onu kaldır' kaldırma onayı bağlamını hafızada korur", () async {
      final memory = _FakeMemory(_slot('telefon', 'Telefon', 'telefon'));
      final writer = _FakeWriter();
      final pipeline = VixrexNluPipeline(
        memory: memory,
        canonicalWriter: writer,
      );

      final sonuc = await pipeline.handle(
        input: 'onu kaldır',
        controller: null,
        onValidate: _validate,
      );

      expect(sonuc.outcome, VixrexNluPipelineOutcome.needsClarification);
      expect(sonuc.message.text, 'Telefon bilgisini kaldırmamı mı istiyorsun?');
      expect(memory.slot?.eylem, 'kaldir');
      expect(writer.writeCount, 0);
    });

    test(
      "kaldırma onayı beklenirken 'evet' gerçek temizleme değerine dönüşür",
      () async {
        final memory = _FakeMemory(
          _slot('telefon', 'Telefon', 'telefon', eylem: 'kaldir'),
        );
        final writer = _FakeWriter();
        final pipeline = VixrexNluPipeline(
          memory: memory,
          canonicalWriter: writer,
        );

        final sonuc = await pipeline.handle(
          input: 'evet',
          controller: null,
          onValidate: _validate,
        );

        expect(sonuc.outcome, VixrexNluPipelineOutcome.handled);
        expect(sonuc.appliedAnahtar, 'telefon');
        expect(sonuc.appliedDeger, isNull);
        expect(sonuc.message.text, 'Telefon bilgisini kaldırdım.');
        expect(writer.lastValues, [null]);
        expect(memory.clearCount, 1);
      },
    );

    test('yalnız kapanış saati tam çalışma saati diye kaydedilmez', () async {
      final memory = _FakeMemory(
        _slot('calismaSaatleri', 'Çalışma Saatleri', 'metin'),
      );
      final writer = _FakeWriter();
      final pipeline = VixrexNluPipeline(
        memory: memory,
        canonicalWriter: writer,
      );

      final sonuc = await pipeline.handle(
        input: 'akşam yedi',
        controller: null,
        onValidate: _validate,
      );

      expect(sonuc.outcome, VixrexNluPipelineOutcome.needsClarification);
      expect(sonuc.message.text, contains('kaçta açıp kaçta kapandığınızı'));
      expect(memory.clearCount, 0);
      expect(writer.writeCount, 0);
    });

    test('yalnız ilçe tam açık adres diye kaydedilmez', () async {
      final memory = _FakeMemory(_slot('adres', 'Açık Adres', 'uzunMetin'));
      final writer = _FakeWriter();
      final pipeline = VixrexNluPipeline(
        memory: memory,
        canonicalWriter: writer,
      );

      final sonuc = await pipeline.handle(
        input: 'Bağcılar',
        controller: null,
        onValidate: _validate,
      );

      expect(sonuc.outcome, VixrexNluPipelineOutcome.needsClarification);
      expect(
        sonuc.message.text,
        'Açık adresi biraz daha ayrıntılı yazar mısın?',
      );
      expect(memory.clearCount, 0);
      expect(writer.writeCount, 0);
    });
  });
}
